### PR TITLE
Expose pending cluster state queue size in node stats

### LINKED ISSUE #13610: Expose pending cluster state queue size in node stats
#13303 and #13062 introduced the notion of pending cluster states and made it easy to know how many cluster states are waiting to be processed by zen discovery. It would be great to expose this information in our node stats for monitoring. Ideally we'd have: total queue size, number of committed cluster states, and number of pending cluster states.

### PR DESCRIPTION
**Pull request by xuzha**
Add stats about the pending cluster state queue: total queue size, number of committed cluster
states, and number of pending cluster states.

closes #13610

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodeStats.java [hunk 1]
```diff
@@ -77,6 +78,9 @@
     @Nullable
     private ScriptStats scriptStats;
 
+    @Nullable
+    private PendingClusterStateStats pendingClusterStateStats;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodeStats.java]: can we call this DiscoveryStats?

File: core/src/main/java/org/elasticsearch/action/search/ClearScrollResponse.java [hunk 2]
```diff
@@ -19,12 +19,12 @@
 
 package org.elasticsearch.action.search;
 
-import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionResponse;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.xcontent.StatusToXContent;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/search/ClearScrollResponse.java]: unneeded?

File: core/src/main/java/org/elasticsearch/discovery/Discovery.java [hunk 3]
```diff
@@ -87,4 +88,10 @@ public FailedToCommitClusterStateException(String msg, Throwable cause, Object..
             super(msg, cause, args);
         }
     }
+
+    /**
+     * @return stats about the discovery
+     */
+    PendingClusterStateStats stats();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/Discovery.java]: can we make this return a DiscoveryStats interface which extends both toXContent and Streamable? ZenDiscovery can return a ZenDiscoveryStats object and other discoveries can decide what they want to expose. This also means we need a method `readStatsFrom(InputStream)` to read a DiscoveryStats object from a stream (only the discovery implementation knows how to deserialize it).

**Comment by bleskes** [comment]
Thanks @xuzha for picking this up. I left some comments. Also, I think it will be great if the xContent output will look like this:

```
...
          "discovery": {
            "cluster_state_queue": {
            "total": 0,
            "pending": 0,
            "committed": 0
         }
```

File: core/src/main/java/org/elasticsearch/discovery/Discovery.java (hunk 3 above)
Comment (xuzha) [inline core/src/main/java/org/elasticsearch/discovery/Discovery.java]: Understood, this suggestion is much better.
I think the only question is how could we expose the info to rest `cat` api, which build table cells need to know what data would be exposed in advance, `RestNodesAction.java`



Hmm, do we need this in `cat` rest api?

File: core/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodeStats.java (hunk 1 above)
Comment (xuzha) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodeStats.java]: Yes, I will refactor this and use a interface.

File: core/src/main/java/org/elasticsearch/discovery/Discovery.java (hunk 3 above)
Comment (xuzha) [inline core/src/main/java/org/elasticsearch/discovery/Discovery.java]: And, I dont't know if I understand this correctly. But look at `NodeStats.java` line 228, I think use a interface here would be really difficult here.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/Discovery.java]: OK. I see - valid point. If you're up to it we can implement a whole registry based on NamedWriteableRegistry and NamedWriteable. Alternatively I'm good with hard coding DiscoveryStats to include whatever ZenDiscovery needs (pending cluster state queue stats only now). We can open it up later if we need to.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/Discovery.java]: Re the cat api - I'm fine if this is not included in them btw - this is a super expert stats.

**Comment by xuzha** [comment]
@bleskes  I pushed another commit. I probably don't get your suggestion. Please take a look and let me know what you think. Thanks again.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import org.elasticsearch.discovery.zen.publish.PendingClusterStateStats;
+
+import java.io.IOException;
+
+public class DiscoveryStats implements Streamable, ToXContent {
+
+    private PendingClusterStateStats queueStats;
+
+    public DiscoveryStats() {
+
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java]: nit cat we explicitly call the other constructor with null? i.e., `this(null)`

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.discovery.zen.publish.PendingClusterStateStats;
+
+import java.io.IOException;
+
+public class DiscoveryStats implements Streamable, ToXContent {
+
+    private PendingClusterStateStats queueStats;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java]: can we mark this as nullable?

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        static final XContentBuilderString DISCOVERY = new XContentBuilderString("discovery");
+    }
+
+    public PendingClusterStateStats getQueueStats() {
+        return queueStats;
+    }
+
+    @Override
+    public String toString() {
+        try {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java]: Why not have the standard to string? A multiline return value is difficult to work with in a debugger...

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStateStats.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        builder.field(Fields.TOTAL, total);
+        builder.field(Fields.PENDING, pending);
+        builder.field(Fields.COMMITTED, committed);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        total = in.readInt();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStateStats.java]: we can use in.readVInt() here, no? it's always non-negative...

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java [hunk 8]
```diff
@@ -283,4 +283,17 @@ public String toString() {
         }
     }
 
+    public PendingClusterStateStats stats() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java]: this needs to be synchronized.

**Comment by bleskes** [comment]
This looks great. I left some minor comments. Can we also add a rest test ?

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStateStats.java (hunk 7 above)
Comment (xuzha) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStateStats.java]: Good call, thx.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java (hunk 6 above)
Comment (xuzha) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java]: Sure, will remove this.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java (hunk 4 above)
Comment (xuzha) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java]: Yes, I removed this DiscoveryStats() constructor

**Comment by xuzha** [comment]
Thanks @bleskes, I updated the PR, please take another look if you have time.

File: core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import org.elasticsearch.discovery.zen.publish.PendingClusterStateStats;
+
+import java.io.IOException;
+
+public class DiscoveryStats implements Streamable, ToXContent {
+
+    @Nullable
+    private PendingClusterStateStats queueStats;
+
+    public DiscoveryStats(PendingClusterStateStats queueStats) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/DiscoveryStats.java]: can we mark the parameter as nullable as well?

File: core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryIT.java [hunk 10]
```diff
@@ -256,4 +261,37 @@ public void testJoinElectedMaster_incompatibleMinVersion() {
...
+                "      \"pending\" : 0,\n" +
+                "      \"committed\" : 0\n" +
+                "    }\n" +
+                "  }\n" +
+                "}";
+
+        Settings nodeSettings = Settings.settingsBuilder()
+                .put("discovery.type", "zen") // <-- To override the local setting if set externally
+                .build();
+        internalCluster().startNode(nodeSettings, Version.V_2_0_0_beta1);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryIT.java]: why do we need this version?

**Comment by bleskes** [comment]
LGTM. Left some minor comments. Feel free to push once addressed. Thanks!

**Comment by xuzha** [comment]
Thanks @bleskes again for the review,  merged.

**Comment by clintongormley** [comment]
@xuzha should we backport to 2.2, 2.x?

**Comment by xuzha** [comment]
@clintongormley hmm, I think we should not. This is PR is based on PendingClusterStatesQueue (https://github.com/elastic/elasticsearch/pull/13303), which is only in master.

**Comment by clintongormley** [comment]
thanks @xuzha

**Pull request merged**